### PR TITLE
Making of NumTok an API for numeral

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** code reorganization

Has an overlay for QuickChick: [#189](https://github.com/QuickChick/QuickChick/pull/189).

Depends on #11688 (merged).

Fixes at least one bug in the parser of decimal numbers for `R`.

We make an API out of the different pieces of code which manipulates numerals, possibly with a sign, a decimal part or an exponent. In particular, we use typing to distinguish when it is about unsigned numbers and when it is about signed numbers, and when it is about integer numbers.

Former type `NumTok.t` is now abstract (internally made of `string`s accepting leading `0`s, and  thousands separated by `_`).

The API is somehow constraining and (therefore) painful. One has to find in the API the right conversions between OCaml's `string` representation of `int`, `bigint`, Coq's abstract `NumTok.SignedNat` and `NumTok.UnsignedNat` representations of signed or unsigned int, Coq's abstract `NumTok.Signed` and `NumTok.Unsigned` representations of signed or unsigned int with decimals and exponents. On the other side, these constraints are probably useful.

- [X] Added / updated test-suite

### CODE REVIEW

File: proofs/logic.ml
Comment (proux01) [inline proofs/logic.ml]: I don't get this change, is it related?

File: interp/numTok.mli
Comment (proux01) [inline interp/numTok.mli]: If I remember well, it is also needed for user defined notations that would contain numerals (c.f. comment below)

https://github.com/coq/coq/blob/18aa9ca60ec9b3d1712276ec0c615dfe54c1a251/interp/constrexpr.ml#L64

File: interp/numTok.mli
Comment (proux01) [inline interp/numTok.mli]: Isn't it `n.p*10^q`?

File: interp/numTok.mli
Comment (proux01) [inline interp/numTok.mli]: `n/10^p`?

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml]: Doesn't this contradict the comment above?

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml]: No more "warnings above"

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml]: `-?` for unsigned numbers?

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml]: This can probably be implemented using `UnsignedNat.compare` above, saving a few lines.

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml]: Don't we need to remove the underscores (like in `to_bigint` just above)?

**Review by proux01 — COMMENTED** [review]
Sounds really nice.

**Comment by proux01** [comment]
The commit message reads "positive natural numbers (may include "_" to separate thousands and leading 0)" but strictly speaking, underscores are allowed everywhere.

File: proofs/logic.ml
Comment (herbelin) [reply proofs/logic.ml]: Sorry, a collision with another PR I was working on at roughly the same time. Fixed

File: interp/numTok.mli
Comment (herbelin) [reply interp/numTok.mli]: I changed the message. Tell me if you consider that it fixes the previous inaccuracy.

File: interp/numTok.mli
Comment (herbelin) [reply interp/numTok.mli]: Right, and similarly below.

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: Indeed, I made a confusion. The `split_on_char '_' s` is actually for the `NumTok.UnsignedNat.t` to `string` translation (where `string` is typically the input syntax for `int_of_string` or `Bigint.of_string).

Moreover, I realize that some of the usage of `NumTok.UnsignedNat.to_string` done outside `NumTok` are rather of the `print` feature. So, I added a `NumTok.UnsignedNat.print` method.

The last use of `NumTok.UnsignedNat.to_string` is in `notation.ml` and I cannot drop the specific treatment of the `_` separator there.

Thanks!

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: I replaced with a reference to `notation.ml`.

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: This is remaining from a previous step of the reworking of `NumTok`, I fixed.

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: Good idea, I did it. (Please check though.)

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: Yes, well spotted.

I'm going tonight to clarify the semantics of the different `to_string` functions.

**Comment by herbelin** [comment]
> The commit message reads "positive natural numbers (may include "_" to separate thousands and leading 0)" but strictly speaking, underscores are allowed everywhere.

Good question. I'm unsure that having `_` is actually very intuitive and I'm wondering whether we should not be more restrictive. I don't know the general feeling.

**Comment by herbelin** [comment]
@proux01: Actually, no need you look now at the change. I'll make tonight a clear diff of the new changes easier to read.

**Comment by herbelin** [comment]
@proux01: Thanks a lot for your careful reading then.

I reorganize a bit the `of_string`/`to_string` according to your remarks. I added more comments in `numTok.mli`. There are now 3 new commits compared to what you read in the morning (one of them to be squashed with the main commit before merging).

The logic for the changes is:
- `of_string`/`to_string` go from OCaml's `int_of_string` or `Bigint.of_string` string format in the `SignedNat`/`UnsignedNat` cases. For OCaml, I realized that `_` is actually supported. We need to remove the `_` for `Bigint.of_string` though (even if `bigint` may eventually disappear in #8743). So maybe overkill to skip `_`, but, at least safer.
- `of_string`/`to_string` goes from OCaml `float_of_string` format in the full `Signed`/`Unsigned` cases.
- new `sprint`/`print` to restore the string as it was parsed, in all types

I have further questions if ever you have an opinion on them:
- Should we parse `+42`, or `+3.5`, the same way as OCaml actually accepts to parse it? This seems quite standard to parse positive s after all, so, maybe should we? I experimentally added a commit for supporting it (in `NumTok`, not yet in the parse).
- Should I complete the API for consistency? For instance, I currently added, even if unused: `of_string`/`to_string`/`of_bigint` in `SignedNat` and `sprint`/`compare` in `UnsignedNat`. I may complete further, e.g. `sprint`/`print` in all categories??
- Should I rename `Signed`/`Unsigned` into `SignedDecimal`/`UnsignedDecimal` to make a stronger difference?

File: interp/numTok.mli
Comment (proux01) [reply interp/numTok.mli]: I would rather move the entire comment from constrexpr.ml (which seems now a bit out of place / inaccurate) to numTok.mli.

File: interp/numTok.ml
Comment (proux01) [reply interp/numTok.ml]: Which compare is this? (I don't see any compare in Unsigned)

File: interp/numTok.ml
Comment (proux01) [inline interp/numTok.ml]: Don't we have to remove potential `_` from `frac` (but not leading zeros)? `UnsignedNat.to_string frac` should do the job.

File: interp/numTok.mli
Comment (proux01) [inline interp/numTok.mli]: s/positive //

File: parsing/g_prim.mlg
Comment (proux01) [inline parsing/g_prim.mlg]: I would rather write "an integer", but my english might be wrong.

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
> I'm unsure that having `_` is actually very intuitive and I'm wondering whether we should not be more restrictive. I don't know the general feeling.

I don't have any strong opinion. The feature seem less pervasive than I initially though. For instance, OCaml and Java both allow `_` everywhere, whereas C or Python 2 don't allow it at all. Python 3 doesn't allows successive `_` but single underscores are still accepted everywhere.

> Should we parse +42, or +3.5,

No opinion, but if it does not harm, certainly nice to have.

> Should I complete the API for consistency?

It's an internal API, probably no need to introduce too much dead code.

> Should I rename Signed/Unsigned into SignedDecimal/UnsignedDecimal

I personally wouldn't. We might for instance want to extend them to hexadecimal at some point in the future.

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: Indeed. This was the wrong `compare` (from `Pervasives`) and, more than that, I completely broke the code. Hopefully fixed.

File: interp/numTok.ml
Comment (herbelin) [reply interp/numTok.ml]: Indeed, this is subtle (and was wrong before). The length needs to be done once `_` are removed.

Here is a bug of 8.11:
```
Require Import Reals.
Check 1.5_%R.
(* 15e-2%R *)
Check 1.5_e2%R.
(* 15%R *)
```

Incidentally, do we really want that numbers with decimal part are printed with `e` in `R` rather than with the decimal part?

File: interp/numTok.mli
Comment (herbelin) [reply interp/numTok.mli]: Indeed.

File: parsing/g_prim.mlg
Comment (herbelin) [reply parsing/g_prim.mlg]: My English is also poor, but I agree with you.

**Comment by herbelin** [comment]
> No opinion, but if it does not harm, certainly nice to have.

I don't think it would be more difficult than parsing `-` (which is not easy)! It is more about whether this is intuitively expected.

> It's an internal API, probably no need to introduce too much dead code.

OK to me.

> >  Should I rename Signed/Unsigned into SignedDecimal/UnsignedDecimal
> 
> I personally wouldn't. 

OK to me.

> We might for instance want to extend them to hexadecimal at some point in the future.

Indeed a nice perspective.

File: interp/numTok.ml
Comment (proux01) [reply interp/numTok.ml]: Damn it, it was introduced in 8.10 and I'm the one to blame.

Printing in `R` could indeed be nicer. I'm adding that on top of my TODO list.

File: interp/numTok.mli
Comment (proux01) [reply interp/numTok.mli]: BTW, shouldn't we take the opportunity to add a test for this, something along the following lines
```Coq
Notation "0" := Prop.
Notation "+0" := O.
Notation "-0" := true.
Notation "00" := (nil, nil).                                                                                                                          
Check 0 : Prop.
Check +0 : nat.
Check -0 : bool.
Check 00 : prod _ _.

Notation "1000" := False.
Notation "1_000" := O.
Check 1000 : Prop.
Check 1_000 : nat.
```

**Review by SkySkimmer — APPROVED** [review]
OK for test suite

**Comment by SkySkimmer** [comment]
>OK for test suite

(the test suite is broken but not the part governed by the test suite codeowner team)

**Comment by herbelin** [comment]
@coq/ci-maintainers: I've seen several time the GitLab pipeline failing without building nothing. Do you know the reason? What should I do?

**Comment by ejgallego** [comment]
> What should I do?

See https://github.com/coq/coq/issues/11320 , you need to cancel a job and then restart it manually :/

**Comment by herbelin** [comment]
If no more comments, I'm going to smash the commits.

**Comment by proux01** [comment]
Ok for me. You may want to add the above suggested test : https://github.com/coq/coq/pull/11703#discussion_r388760268

**Comment by herbelin** [comment]
> You may want to add the above suggested test

Indeed. Sorry about missing them. Done, and squashed.

Note: I reverted the commit `Notation of the form "-x" should be written "- x"` as this does not fix all (factorization) issues with negative numbers; I will address these issues in another PR.

By the way, if ever you are interested to be maintainers for such kind of files, my understanding is that you can get in touch with @zimmi48.

File: interp/numTok.mli
Comment (proux01) [inline interp/numTok.mli]: BTW, I wonder whether this isn't actually `n*10^p`, in which case the name might be a bit strange. But that's just a detail.

Comment (herbelin) [reply interp/numTok.mli]: Yes, and I don't know why I started to be convinced it was a `/`. Thanks, it is not a detail after all. What about calling them `to_bigint_and_exponent` and `of_bigint_and_exponent` then?

Comment (proux01) [reply interp/numTok.mli]: sounds good

**Review by ppedrot — APPROVED** [review]
LGTM. Maybe the NumTok API could be made slightly more palatable using GADTs, but that is not necessary in this PR.

**Comment by ppedrot** [comment]
@herbelin this has been sitting for a long time, I'll merge it when you rebase and after the CI.

**Comment by herbelin** [comment]
I rebased. The VST failure seems specific to VST. The failures in Equations/Elpi/rewriter are due to the overlays of #11731 not being merged yet. Note the existence of an overlay for QuickChick.

**Pull request merged**
